pkg/syncer/cnsoperator/controller/cnsnodevmbatchattachment/cnsnodevmbatchattachment_controller.go

```diff
@@ -0,0 +1,797 @@
+
+		msg := fmt.Sprintf("ReconcileCnsNodeVmBatchAttachment: Successfully processed instance %s "+
+			"in namespace %q.", request.Name, request.Namespace)
+		recordEvent(batchAttachCtx, r, instance, v1.EventTypeNormal, msg)
+		return reconcile.Result{}, nil
+	}
+
+	reconcileLog.Infof("Reconcile for CnsNodeVmBatchAttachment request: %q completed.", request.NamespacedName)
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileCnsNodeVmBatchAttachment) reconcileInstanceWithDeletionTimestamp(ctx context.Context,
+	instance *cnsnodevmbatchattachmentv1alpha1.CnsNodeVmBatchAttachment,
+	volumesToDetach map[string]string,
+	nodeVM *cnsvsphere.VirtualMachine,
+	pvcToVolumeId map[string]string) error {
+	log := logger.GetLogger(ctx)
+
+	err := r.detachVolumes(ctx, nodeVM, instance, volumesToDetach)
+	if err != nil {
+		instance.Status.Error = err.Error()
+		log.Errorf("failed to detach all volumes. Err: +v", err)
+		return err
+	}
+
+	// CR is being deleted and all volumes were detached successfully
+	removeFinalizerFromCRDInstance(ctx, instance)
+	return nil
+}
+
+// reconcileInstanceWithoutDeletionTimestamp calls batch attach a well as individual detach calls
+// for volumes present in volumesToAttach and volumesToDetach lists.
+func (r *ReconcileCnsNodeVmBatchAttachment) reconcileInstanceWithoutDeletionTimestamp(ctx context.Context,
+	instance *cnsnodevmbatchattachmentv1alpha1.CnsNodeVmBatchAttachment,
+	volumesToAttach map[string]string, volumesToDetach map[string]string,
+	nodeVM *cnsvsphere.VirtualMachine, pvcToVolumeId map[string]string) error {
+	log := logger.GetLogger(ctx)
+
+	// Call batch attach for volumes which need to be attached.
+	if len(volumesToAttach) != 0 {
+		err := r.attachVolumes(ctx, nodeVM, pvcToVolumeId, instance, volumesToAttach)
+		if err != nil {
+			instance.Status.Error = err.Error()
+			log.Errorf("failed to attach all volumes. Err: %+v", err)
+			return err
+		}
+	}
+
+	// Call detach if there are some volumes which need to be detached.
+	if len(volumesToDetach) != 0 {
+		err := r.detachVolumes(ctx, nodeVM, instance, volumesToDetach)
+		if err != nil {
+			instance.Status.Error = err.Error()
+			log.Errorf("failed to detach all volumes. Err: +v", err)
+			return err
+		}
+		log.Infof("Successfully detached all volumes %+v", volumesToDetach)
+	}
+
+	return nil
+}
+
+// isInstanceProcessed finds the VM on vCenter to find the list of FCDs attached to it.
+// It then takes a diff of those FCDs and the ones in spec to find out
+// which volumes need to be attached and which ones need to be detached.
+//
+// If VM retrieval from the VC fails,
+// then fail this operation as it is invalid.
+//
+// isInstanceProcessed returns true if no attach or detach is required.
+// It also returns list of volumes to attach, list of volumes to detach,
+// VM object and error.
+func (r *ReconcileCnsNodeVmBatchAttachment) isInstanceProcessed(ctx context.Context,
+	volumeIdToPvc map[string]string, pvcToVolumeId map[string]string,
```

Comment:
Can we avoid suppling volumeIdToPvc and pvcToVolumeId to isInstanceProcessed.

Use Informer cache.